osu.Game.Rulesets.Taiko/UI/DrumTouchInputArea.cs

```diff
@@ -0,0 +1,141 @@
+
+        // A map of (Finger Index OnTouchDown -> Which Taiko action was pressed), so that the corresponding action can be released OnTouchUp is released even if the touch position moved
+        private Dictionary<TouchSource, TaikoAction> touchActions = new Dictionary<TouchSource, TaikoAction>(Enum.GetNames(typeof(TouchSource)).Length);
+
+        private Playfield playfield;
+
+        public DrumTouchInputArea(Playfield playfield) {
+            this.playfield = playfield;
+
+            RelativeSizeAxes = Axes.Both;
+            RelativePositionAxes = Axes.Both;
+            Children = new Drawable[]
+            {
+                new Box() {
+                    Alpha = 0.0f,
+                    Colour = new OsuColour().Blue,
+
+                    RelativeSizeAxes = Axes.Both,
+                    RelativePositionAxes = Axes.Both,
+                },
+                new Container() {
+                    RelativeSizeAxes = Axes.Both,
+                    RelativePositionAxes = Axes.Both,
+
+                    Anchor = Anchor.BottomCentre,
+                    Origin = Anchor.BottomCentre,
+
+                    Children = new Drawable[]
+                    {
+                        touchInputDrum = new InputDrum(playfield.HitObjectContainer) {
+                            Anchor = Anchor.Centre,
+                            Origin = Anchor.Centre,
+                        },
+                    }
+                },
+            };
+
+        }
+        protected override void LoadComplete()
+        {
+            keyBindingContainer = taikoInputManager?.KeyBindingContainer;
+
+            Padding = new MarginPadding {
+                Top =  playfield.ScreenSpaceDrawQuad.BottomLeft.Y,
+                Bottom = -DrawHeight * overhangPercent,
+            };
+        }
+
+        protected override bool OnMouseDown(MouseDownEvent e)
+        {
+            mouseAction = getTaikoActionFromInput(e.ScreenSpaceMouseDownPosition);
+            keyBindingContainer?.TriggerPressed(mouseAction);
+            return base.OnMouseDown(e);
+        }
+
+        protected override void OnMouseUp(MouseUpEvent e)
+        {
+            keyBindingContainer?.TriggerReleased(mouseAction);
+            base.OnMouseUp(e);
+        }
+
+        protected override bool OnTouchDown(TouchDownEvent e)
+        {
+            TaikoAction taikoAction = getTaikoActionFromInput(e.ScreenSpaceTouchDownPosition);
+            if (touchActions.ContainsKey(e.Touch.Source)) {
+                touchActions[e.Touch.Source] = taikoAction;
+            }
+            else {
+                touchActions.Add(e.Touch.Source, taikoAction);
+            }
+            keyBindingContainer?.TriggerPressed(touchActions[e.Touch.Source]);
+
+            return base.OnTouchDown(e);
+        }
+
+        protected override void OnTouchUp(TouchUpEvent e)
+        {
+            keyBindingContainer?.TriggerReleased(touchActions[e.Touch.Source]);
```

Comment:
99% sure this also won't work because the `TouchDown` isn't handled.

Also you should probably remove the action from the dictionary here after consumption.

Reply:
Not really sure what you mean about `TouchDown` not being handled?

As for not removing the action from the  dictionary, I was thinking that I should be using the dictionary as if it were fixed size, since I was thinking calling `remove` may result in the dictionary's memory allocation being resized - which could mean it's not contiguous in memory which might be less performant? Should use an array of pairs if I want to structure it that way? Or am I just overcomplicating this in general and I should just let the dynamic allocation happen?

Reply:
You need to return `true`. You'll notice this if you actually test on a touch device. That would be the minimum requirement for pull requesting touch support for a game, I'd hope.

Reply:
Ah gotcha.

And for the performance - sounds fine to me. I'll call remove.

Reply:
Made recommended changes as of 317869078f470889330702b7b99b6146569b4f75